search/materialize.go

```diff
@@ -275,35 +270,79 @@ func totalRows(rr []RowRange) int64 {
 	return res
 }
 
-func (m *Materializer) materializeChunks(ctx context.Context, rgi int, mint, maxt int64, rr []RowRange) ([][]chunks.Meta, error) {
+func (m *Materializer) materializeChunks(rgi int, mint, maxt int64, rr []RowRange) chunkIterableSet {
 	minDataCol := m.s.DataColumIdx(mint)
 	maxDataCol := m.s.DataColumIdx(maxt)
 	rg := m.b.ChunksFile().RowGroups()[rgi]
-	r := make([][]chunks.Meta, totalRows(rr))
 
-	for i := minDataCol; i <= min(maxDataCol, len(m.dataColToIndex)-1); i++ {
-		values, err := m.materializeColumn(ctx, m.b.ChunksFile(), rg, rg.ColumnChunks()[m.dataColToIndex[i]], rr)
-		if err != nil {
-			return r, err
-		}
+	dataColCount := min(maxDataCol, len(m.dataColToIndex)-1) - minDataCol + 1
 
-		for vi, value := range values {
-			chks, err := m.d.Decode(value.ByteArray(), mint, maxt)
-			if err != nil {
-				return r, errors.Wrap(err, "failed to decode chunks")
-			}
-			r[vi] = append(r[vi], chks...)
-		}
+	if dataColCount <= 0 {
+		return &repeatEmptyChunkIterableSet{count: totalRows(rr)}
 	}
 
-	return r, nil
+	result := &chunksIteratorIterator{
+		d:    m.d,
+		mint: mint,
+		maxt: maxt,
+	}
+
+	for colIdx := range dataColCount {
+		iter := m.materializeColumn(m.ctx, m.b.ChunksFile(), rg, rg.ColumnChunks()[m.dataColToIndex[minDataCol+colIdx]], rr)
```

Comment:
Wondering what happens if the context gets cancelled and the iterator was not released.